Add plots

LukeFiddy opened the pull request:
adding plot functionality and integrating into cli

Commit:
add plto classes

Commit:
add beamline scan plot class

Commit:
change formatting

Commit:
change formatting

Commit:
change formatting

Commit:
add matplotlib dependency

Commit:
integrating plotting into cli

Commit:
add asve line to mirror surface plot

Commit:
fix pyplot typo in imports

Commit:
add print statements to generate_plots

Commit:
add test for generate-plots cli command

Commit:
change pencil beam scan plot to plot against slit positions

Commit:
add tests for plot classes

Commit:
add type:ignore for pandas

Commit:
formatting

Commit:
add comments

Commit:
add test for Plot.save_plot

Commit:
change save+plot test

dan-fernandes commented on src/bimorph_mirror_analysis/__main__.py:
```diff
@@ -54,12 +58,76 @@ def calculate_voltages(
     )
 
 
+def calculate_optimal_voltages(file_path: str) -> np.typing.NDArray[np.float64]:
```
Could: use the pathlib library instead of strings to represent paths. Might be overkill, but could simplify the validation and  logic you're doing in this file to manipulate strings representing paths. Also gives a nicer function signature eg: `def read_bluesky_plan_output(filepath: str): ..` vs `def read_bluesky_plan_output(filepath: Path): ..`.

LukeFiddy replied:
added a ticket to do this. will implement once core functionality for the in-lab tests are completed

dan-fernandes commented on src/bimorph_mirror_analysis/__main__.py:
```diff
@@ -54,12 +58,76 @@ def calculate_voltages(
     )
 
 
+def calculate_optimal_voltages(file_path: str) -> np.typing.NDArray[np.float64]:
```
Should: Add docstring

LukeFiddy replied:
added

LukeFiddy commented:
Must add traces for both vltage calculations before merging. This requires the completion of https://github.com/DiamondLightSource/bimorph-mirror-analysis/issues/31

Commit:
add docstring to calculate_optimal_voltages

Commit:
remove incrememnt as input to generate-plots cli command

Commit:
add function to test whether voltages fit constraints

Commit:
add selective method for voltage calculation to cli

Commit:
add flag to allow cli inputs containing a dash

Commit:
change tests to use added calculate-voltages arguments

now uses the maximum consecutive difference and the voltage range arguments

Commit:
add predicted centroid traces to mirror surface plot created by genereate-plots command

Commit:
update test for new cli generate-plots command inputs

Commit:
remove unused variable

Commit:
add initial voltages files

Commit:
change actuator data fixture to accomodate initial voltages file

Commit:
change script tests to accomodate addition of initial voltages to fixture

Commit:
change read_bluesky_plan_output to accomodate non 0 baseline scan indices

Commit:
add baseline voltage scan as option to cli for calculating voltages

Commit:
incorporate basline voltage scan index option for cli into tests

Commit:
remove print statements

Commit:
fix handling of negative baseline voltage san indices in read_bluesky_plan_output

Commit:
add test for negative index for read_bluesky_plan_output

Commit:
fix calculation of voltages with restraints

previously the restraints were applied to the voltage changes, not the voltages themselves

Commit:
fix cli test for plots to use fixed voltage calculation with restraints

the initial voltages were not mocked properly, was just returning an int because this previously did not throw an error

Commit:
add test to ensure restrained voltage calcultion is used if the optimal does not fict constraints

Commit:
adjust tests for the constrained calculations

Commit:
adjust cli commands to keep initial_voltages, change helper function to accomodate

dan-fernandes commented on src/bimorph_mirror_analysis/__main__.py:
```diff
@@ -15,18 +24,28 @@
+@app.command(name=None, context_settings={"ignore_unknown_options": True})
 def calculate_voltages(
     file_path: str = typer.Argument(help="The path to the csv file to be read."),
+    voltage_range: tuple[int, int] = typer.Argument(
```
Could: make it more obvious that this expects two integers. On first glance the help generated from this implies it is only one argument, as running `--help` does not show that it expects a tuple.

LukeFiddy replied:
Have adjusted the help

dan-fernandes commented on src/bimorph_mirror_analysis/__main__.py:
```diff
@@ -54,12 +78,135 @@ def calculate_voltages(
+) -> np.typing.NDArray[np.float64]:
+    """Calculate the optimal voltages for the bimorph mirror actuators.
+
+    Args:
```
Should: Fix docstring args

LukeFiddy replied:
fixed

dan-fernandes commented on src/bimorph_mirror_analysis/__main__.py:
```diff
@@ -54,12 +78,135 @@ def calculate_voltages(
+    ):
+        return optimal_voltages
+
+    else:
```
Could: Remove unnecessary `else`. If the `if` branch executes, the function will `return` and stop execution.

LukeFiddy replied:
imo it makes it more obvious what is happening if the else block is kept

dan-fernandes commented on src/bimorph_mirror_analysis/__main__.py:
```diff
@@ -54,12 +78,135 @@ def calculate_voltages(
+            voltage_range,
+            max_consecutive_voltage_difference,
+            baseline_voltage_scan=baseline_voltage_scan,
+        )  # type: ignore
```
Nit: double check the `# type: ignore` here and on 121 are necessary. My type checker didn't complain, but it might just have not updated.

LukeFiddy replied:
you're right, they were left over from before some changes i think. have removed them now

dan-fernandes commented on src/bimorph_mirror_analysis/maths.py:
```diff
@@ -144,18 +144,24 @@ class Constraint(TypedDict):
-            return max_diff - abs(voltages[idx] - voltages[idx + 1])
+            # voltages has an extra element at the start so idx always needs to be +1
+            return max_diff - abs(
+                (initial_voltages[idx] + voltages[idx + 1])
```
Could: It took me a second to realise these are voltage adjustments, which is why they're being added to the initial voltages. (Is this right?) You could consider renaming the parameters to `voltage_adjustments`.

LukeFiddy replied:
Done

dan-fernandes commented on src/bimorph_mirror_analysis/read_file.py:
```diff
@@ -21,10 +23,25 @@ def read_bluesky_plan_output(
+
+    diff_in_voltages: np.typing.NDArray[np.float64] = other_voltages - initial_voltages  # type: ignore
+    max_diff, min_diff = np.max(diff_in_voltages), np.min(diff_in_voltages)  # type: ignore
+    if abs(max_diff) > abs(min_diff):
```
Could: `voltage_increment = max(abs(max_diff), abs(min_diff))`

LukeFiddy replied:
this method loses the sign of the increment

Commit:
remove unnecesary interaction matrix input to the generate_minimize_constraints function

Commit:
rename voltage adjustments variable

Commit:
adjust help for voltage range cli argument

now explains that 2 values separated by a space are required

Commit:
add docstring arguments for calculate_optimal_voltages

Commit:
remove unneccesary type: ignores

dan-fernandes reviewed (approved):
LGTM
